### PR TITLE
[Play] - Card Answers

### PR DESCRIPTION
**Pull request by drewjhart**
Adds `RoundTextIcon`, `Card`, `CardAnswer` components and inserts them into `GameInProgress`. 

**`RoundTextIcon`**
Function:
- This is the answer button that players use to select their answer

Layout:
- Flexes to expand to parent container, multiple states depending on if the answer is selected, if the answer is correct and if the player has submitted an answer (see storybook)

Props:
- `answerStatus` if the answer is correct, selected or neither (I think I'll need to refine the state later in development)
- `index` used to track currently selected answer
- `answerText` populates button with answer text
- `isSubmitted` boolean to determine if player has submitted answer
- `handleSelectAnswer` callback to handle when a player presses the RoundTextIcon (ie they have selected an answer)

Storybook:
https://64220218d457f84ba4546eb1-vpztfswifn.chromatic.com/?path=/story/design-system-2-molecules-roundtexticon--default-answer

**`Card`**
Function:
- Card container that wraps cardContent to provide formatting

Layout:
- located per flex of parent, border radius and shadow provided. 

Props:
-`extraStyle` - prop to handle additional styling as req'd
- `children` - passes down cardContent
- `headerTitle` - identifying string for each card

Storybook:
N/A

**`CardAnswer`**
Function:
- Card contents for player answering in `CHOOSE_CORRECT_ANSWER` and `CHOOSE_TRICKIEST_ANSWER` screens

Layout:
- content populated in `Card` container

Props:
-`answers` - game answers for `.map` distribution to `RoundTextIcon
- `isSubmitted` - tracks if a player has submitted their answer (state management happening in `GameSessionContainer`). Passed down to `RoundTextIcon`
- `currentState` - used to determine text of Card answer
- `selectedAnswer` - passed down to `RoundTextIcon` to determine formatting
- `handleSelectAnswer` - callback from and passed down to `RoundTextIcon`

Storybook:
https://64220218d457f84ba4546eb1-vpztfswifn.chromatic.com/?path=/story/design-system-3-organisms-cardanswer--correct-answer

@amarax: given how we are managing state, it's easier to see the function of `CardAnswer` inputs using this W.I.P `GameInSession` story (I've built this just so you can see button presses. Layout is in -progress):
https://64220218d457f84ba4546eb1-vpztfswifn.chromatic.com/?path=/story/design-system-4-pages-gameinprogress--test-story-one

### CODE REVIEW

**Comment by drewjhart** [comment]
Hi @maniramezan and @amarax,
I'm currently working on individual components and then merging them into [PR575](https://github.com/rightoneducation/righton-app/pull/575). This PR provides `RoundTextIcon`, `Card` and `CardAnswer` components as well as a demo `GameInProgress` story. For more information, see the above description. 

Ideally, I'd like to get this approved and then get PR575 approved as well. I feel like that PR is ballooning quickly as we integrated these smaller PR's into it. In the meantime, I'm working on laying out the `GameInProgress` screen.

File: play/src/components/RoundTextIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    'correct': classes.buttonCorrect,
+    'selected': classes.buttonSelected,
+  }
+ 
+  return (
+    <div className={classes.container}>
+      <Button
+        className={`${classes.buttonBase} ${styleMap[answerStatus]}`}
+        onClick={() => handleSelectAnswer(index)}
+        disabled={isSubmitted || answerStatus === 'correct'}
```
Comment (maniramezan) [inline play/src/components/RoundTextIcon.tsx:43]: Instead of repeating `'correct'` string in this file, probably better to create a local variable to assign it to `'correct'` and then use that. Even better, use enum to define answer status options and then you can even assign strings to each case. That way, it's both strongly typed, and also has a value associated to each case. [This doc](https://www.typescriptlang.org/docs/handbook/enums.html) has an example for it. I believe I'm also using this mechanism in networking for `GameSessionState`

File: play/src/components/RoundTextIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        variant="text"
+      >
+        <div className={`${classes.letterText} ${isSubmitted || answerStatus === 'correct' ? classes.letterTextSubmitted : ''}`}>  
+          {`${String.fromCharCode(letterCode)}.`} 
+        </div>
+        <div className={classes.answerText}>  {answerText} </div>
```
Comment (maniramezan) [inline play/src/components/RoundTextIcon.tsx:49]: Any reason for extra spaces before and after?

File: play/src/pages/GameInProgress.tsx [hunk 3]
```diff
@@ -1,32 +1,67 @@
...
+import CardAnswer from '../components/CardAnswer';
 import FooterContent from '../components/FooterContent';
 
 interface GameInProgressProps {
   teams?: ITeam[];
   id: string;
   currentState: GameSessionState;
   teamAvatar: number;
+  questions: IQuestion[];
+  currentQuestionIndex?: number | null;
```
Comment (maniramezan) [inline play/src/pages/GameInProgress.tsx:16]: Do you need both `?` and `null`?

File: play/src/pages/GameInProgress.tsx [hunk 4]
```diff
@@ -1,32 +1,67 @@
...
+    return {
+        id: uuidv4(),
+        text: choice.text,
+        isCorrectAnswer: choice.isAnswer,
+    }
+  })
 
   const handleTimerIsFinished = () => {
-    console.log('finished');
+    setTimerIsPaused(true);
```
Comment (maniramezan) [inline play/src/pages/GameInProgress.tsx:50]: Just a note, have tested this when user switches tabs as well?

**Review by maniramezan — COMMENTED** [review]
I'm not CSS expert so didn't look much into those. Not even sure if it makes sense to use some sort of CSS framework or not so leave it to you and Ryan for that part.

Main comments are around some code structures and suggestions for future-proofing.

Great job also on breaking it into smaller components

File: play/src/components/ButtonSubmitAnswer.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    borderRadius: '22px',
+    textTransform: 'none',
+  },
+  buttonStyleSelected: {
+    background: 'linear-gradient(90deg, #159EFA 0%, #19BCFB 100%)',
+    boxShadow: '0px 5px 22px rgba(71, 217, 255, 0.3)'
+  },
+  buttonText: {
+    color: '#FFFFFF',
+    fontFamily: 'Karla',
```
[the remark this answers is not in the document]
Comment (amarax) [reply play/src/components/ButtonSubmitAnswer.tsx:45]: I agree with this, and this is part of the todo for the designs in Figma. Designers refer to these colours/styles as [tokens](https://m3.material.io/foundations/design-tokens/overview).

**Comment by amarax** [comment]
Notes on the component ([Figma link](https://www.figma.com/file/PadzU8GqyhFLCjIt7rQkpE/RightOn-Apps?node-id=12363-382671&t=0dKWaRr4WCgxzway-4)). In general I will list my proposed changes in order of highest to lowest importance.

Important changes that affect usability:
- If the answer text is long, it should look like this https://www.figma.com/file/PadzU8GqyhFLCjIt7rQkpE/RightOn-Apps?node-id=13583-157592&t=0dKWaRr4WCgxzway-4. Specifically: left-aligned, the component grows with the text. We will need to consider this especially if the answer includes inline images in future.

Visual changes (nice to have)
- There is a 1px border for before/after submission. The border increases to 2px only when it is selected before submission. This makes it clear that they are interactive during that game phase, or were previously interactive.
- There is no more border when the option index is used for reviews. This makes it clear that they are not to be interacted with during that game phase.
- The option index is formatted differently, it is slightly smaller than the answer, and has no period character after. It should be aligned such that its baseline is the same as the first line of the answer text.

**Comment by amarax** [comment]
Recommendations for how to organise Storybook:

- Instead of creating a story for each "state" of the component, I suggest just a story for each game state, and update the allowed controls to reflect what's possible/not possible during the game states. This simplifies the number of stories you need to create, and the designer should use the controls to try out different component states instead.
- In addition, this helps to communicate how the game state will impose different constraints on what the component can or cannot do. It's a good form of additional documentation, but will also allow (in future) automated systems to correctly permute the possible states for the component, correctly excluding "impossible" states.
- Also include a default state that isn't tied to a game state, which exposes all the controls that are allowed for the component. This will allow the designer to figure out if it's a suitable component to use for a new game state that hasn't been built.

(The following are examples to illustrate what I mean by the above, it doesn't mean that this is what needs to be done! These are just suggestions based on what I've found to work.)

For example:

**SmallButton**
- Default
   - All controls exposed
   - Labels for before and after submission are exposed
- Answer not Submitted
   - `isSubmitted` is not exposed
   - `isSelected` can be toggled (because this is something that can happen while the user is answering), and the designer can preview the transition for a rapid switch.
- Answer Submitted
   - Neither `isSubmitted` nor `isSelected` is exposed.

In this case, the surrounding "game state" that the SmallButton is concerned with is just whether the answer has been submitted or not. Whether the answer has been selected is arguable if it's a "game state" or not; I personally feel (loosely-held opinion) that because the user stays on the "same screen", won't feel like a "game state" change to them, while submitting their answer is much more significant and and the screen will change a lot more, hence that's definitely a "game state" change.

**RoundTextIcon**
- Default
   - All controls are exposed.
- Answering Phase
   - "correct" `answerStatus` is not exposed, since it should never be revealed what the correct answer is.
- Reviewing Phase
   - `isSubmitted` is not exposed, since during the review it is always considered submitted.
   
For this, Answering and Reviewing are two conceptual types of phases we have in our game. To the user, they are different mindsets that they have to be in, and hence the component should support that. On the other hand, I can use the Storybook Controls to change the internal state of the component to play around with the interactivity.


To summarise:
- I recommend to organise Storybook stories to fit conceptual "game state", to reduce the number of stories for each component.
- This also acts as documentation for what's allowed during each "game state".
- Include a "default" story to expose all controls.

File: play/src/components/ButtonSubmitAnswer.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (drewjhart) [reply play/src/components/ButtonSubmitAnswer.tsx]: Great call, a `Theme` file has been added at `lib\Theme.tsx` that handles colors and fonts throughout the app. I just pulled these off Figma but they can be adjusted when needed.

File: play/src/components/ButtonSubmitAnswer.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  const classes = useStyles();
+  const buttonText = isSubmitted ? "Answer Submitted" : "Submit Answer";
+
+  return (
+    <Button 
+      className={`${classes.buttonStyleBase} ${(isSelected && !isSubmitted) ? classes.buttonStyleSelected : null}`}
+      onClick={() => { 
+        handleSubmitAnswer(true);
+      }}
+      disabled={!isSelected || isSubmitted}
```
[the remark this answers is not in the document]
Comment (drewjhart) [reply play/src/components/ButtonSubmitAnswer.tsx]: This is really helpful. I've added ESLint and prettier to the project. I've also added a `readme.md` file and you can find more information on the ESLint/prettier setup/rules [there](https://github.com/rightoneducation/righton-app/tree/drew-hart--play-cardanswers/play)

File: play/src/components/CardAnswer.tsx [hunk 7]
```diff
@@ -0,0 +1,89 @@
+import React from 'react';
+import Card from './Card';
+import { makeStyles } from '@mui/styles';
+import RoundTextIcon from './RoundTextIcon';
+import ButtonSubmitAnswer from './ButtonSubmitAnswer';
+import { isNullOrUndefined, GameSessionState } from '@righton/networking';
+
+interface CardAnswerProps {
+  answers: {text: string, isCorrectAnswer: boolean}[] | undefined;
```
[the remark this answers is not in the document]
Comment (drewjhart) [reply play/src/components/CardAnswer.tsx]: Honestly, yeah that makes a lot of sense. I'd like to do that in a separate PR after I finalize the first page's UI

File: play/src/components/CardAnswer.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  isSubmitted,
+  handleSubmitAnswer,
+  currentState,
+  selectedAnswer,
+  handleSelectAnswer
+} : CardAnswerProps) {
+  const classes = useStyles();
+  const correctText = (
+    <> 
+      Choose the <div className={classes.titleCorrectText}> correct answer </div>
```
[the remark this answers is not in the document]
Comment (drewjhart) [reply play/src/components/CardAnswer.tsx]: Spacing has been handled now through prettier

File: play/src/components/RoundTextIcon.tsx [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  },
+  buttonBase: {
+      width: '100%', 
+      height: '100%',
+      borderRadius: '22px',
+      flexDirection: 'row',
+      justifyContent: 'flex-start',
+      alignItems: 'center',
+      textTransform: 'none',
+  },
```
[the remark this answers is not in the document]
Comment (drewjhart) [reply play/src/components/RoundTextIcon.tsx]: formatting through prettier

File: play/src/components/RoundTextIcon.tsx (hunk 1 above)
Comment (drewjhart) [reply play/src/components/RoundTextIcon.tsx]: Nice, lin\PlayModels.tsx has been created with an enum for this purpose

File: play/src/components/RoundTextIcon.tsx (hunk 2 above)
Comment (drewjhart) [reply play/src/components/RoundTextIcon.tsx]: fixed via prettier

File: play/src/pages/GameInProgress.tsx (hunk 4 above)
Comment (drewjhart) [reply play/src/pages/GameInProgress.tsx:114]: good point, the timer currently doesn't pause for anything. I'll need to add that in but I'd like to stay focused on broader UI and then fine tune that functionality in a subsequent pull request. To make sure I'm not just kicking the can down the road, I've added a TODO section to the readme to chart these items (with links to new Github issues)

**Comment by drewjhart** [comment]
> Notes on the component ([Figma link](https://www.figma.com/file/PadzU8GqyhFLCjIt7rQkpE/RightOn-Apps?node-id=12363-382671&t=0dKWaRr4WCgxzway-4)). In general I will list my proposed changes in order of highest to lowest importance.
> 
> Important changes that affect usability:
> 
> * If the answer text is long, it should look like this https://www.figma.com/file/PadzU8GqyhFLCjIt7rQkpE/RightOn-Apps?node-id=13583-157592&t=0dKWaRr4WCgxzway-4. Specifically: left-aligned, the component grows with the text. We will need to consider this especially if the answer includes inline images in future.
> 
> Visual changes (nice to have)
> 
> * There is a 1px border for before/after submission. The border increases to 2px only when it is selected before submission. This makes it clear that they are interactive during that game phase, or were previously interactive.
> * There is no more border when the option index is used for reviews. This makes it clear that they are not to be interacted with during that game phase.
> * The option index is formatted differently, it is slightly smaller than the answer, and has no period character after. It should be aligned such that its baseline is the same as the first line of the answer text.

I think these are closed off now (minus the review side comment, which will be handled more on that screen). Here's an update:

image

**Comment by drewjhart** [comment]
> Recommendations for how to organise Storybook:
> 
> * Instead of creating a story for each "state" of the component, I suggest just a story for each game state, and update the allowed controls to reflect what's possible/not possible during the game states. This simplifies the number of stories you need to create, and the designer should use the controls to try out different component states instead.
> * In addition, this helps to communicate how the game state will impose different constraints on what the component can or cannot do. It's a good form of additional documentation, but will also allow (in future) automated systems to correctly permute the possible states for the component, correctly excluding "impossible" states.
> * Also include a default state that isn't tied to a game state, which exposes all the controls that are allowed for the component. This will allow the designer to figure out if it's a suitable component to use for a new game state that hasn't been built.
> 
> (The following are examples to illustrate what I mean by the above, it doesn't mean that this is what needs to be done! These are just suggestions based on what I've found to work.)
> 
> For example:
> 
> **SmallButton**
> 
> * Default
>   
>   * All controls exposed
>   * Labels for before and after submission are exposed
> * Answer not Submitted
>   
>   * `isSubmitted` is not exposed
>   * `isSelected` can be toggled (because this is something that can happen while the user is answering), and the designer can preview the transition for a rapid switch.
> * Answer Submitted
>   
>   * Neither `isSubmitted` nor `isSelected` is exposed.
> 
> In this case, the surrounding "game state" that the SmallButton is concerned with is just whether the answer has been submitted or not. Whether the answer has been selected is arguable if it's a "game state" or not; I personally feel (loosely-held opinion) that because the user stays on the "same screen", won't feel like a "game state" change to them, while submitting their answer is much more significant and and the screen will change a lot more, hence that's definitely a "game state" change.
> 
> **RoundTextIcon**
> 
> * Default
>   
>   * All controls are exposed.
> * Answering Phase
>   
>   * "correct" `answerStatus` is not exposed, since it should never be revealed what the correct answer is.
> * Reviewing Phase
>   
>   * `isSubmitted` is not exposed, since during the review it is always considered submitted.
> 
> For this, Answering and Reviewing are two conceptual types of phases we have in our game. To the user, they are different mindsets that they have to be in, and hence the component should support that. On the other hand, I can use the Storybook Controls to change the internal state of the component to play around with the interactivity.
> 
> To summarise:
> 
> * I recommend to organise Storybook stories to fit conceptual "game state", to reduce the number of stories for each component.
> * This also acts as documentation for what's allowed during each "game state".
> * Include a "default" story to expose all controls.

Great comments. We spoke about this at the U/X coordination meeting on April 10th. Next steps:
1. Dev to provide only 'default' stories for Atom and Organism scale components and a bit more clarity on Stories generally so they are communicating more the states the user will encounter rather than a cross section of all the different states the props can be in.
